sdk/resourcemanager/Azure.ResourceManager/src/Resources/Custom/ProviderCollection.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            string version;
+            Dictionary<string, string> resourceVersions;
+            if (!ClientOptions.ResourceApiVersions.TryGetValue(resourceType.Namespace, out resourceVersions))
+            {
+                resourceVersions = LoadResourceVersionsFromApi(resourceType.Namespace, cancellationToken);
+                ClientOptions.ResourceApiVersions.TryAdd(resourceType.Namespace, resourceVersions);
+            }
+            if (!resourceVersions.TryGetValue(resourceType.Type, out version))
+            {
+                throw new InvalidOperationException($"Invalid resource type {resourceType}");
+            }
+            return version;
+        }
+
+        internal async ValueTask<string> TryGetApiVersionAsync(ResourceType resourceType, CancellationToken cancellationToken = default)
+        {
+            string version;
+            Dictionary<string, string> resourceVersions;
+            if (!ClientOptions.ResourceApiVersions.TryGetValue(resourceType.Namespace, out resourceVersions))
+            {
+                resourceVersions = await LoadResourceVersionsFromApiAsync(resourceType.Namespace, cancellationToken).ConfigureAwait(false);
+                ClientOptions.ResourceApiVersions.TryAdd(resourceType.Namespace, resourceVersions);
+            }
+            if (!resourceVersions.TryGetValue(resourceType.Type, out version))
+            {
+                throw new InvalidOperationException($"Invalid resource type {resourceType}");
+            }
+            return version;
+        }
+
+        private Dictionary<string, string> LoadResourceVersionsFromApi(string resourceNamespace, CancellationToken cancellationToken = default)
+        {
+            Provider results = Get(resourceNamespace, cancellationToken: cancellationToken);
+            return GetVersionsFromResult(results);
+        }
+
+        private async Task<Dictionary<string, string>> LoadResourceVersionsFromApiAsync(string resourceNamespace, CancellationToken cancellationToken = default)
+        {
+            Provider results = await GetAsync(resourceNamespace, cancellationToken: cancellationToken).ConfigureAwait(false);
+            return GetVersionsFromResult(results);
+        }
+
+        private static Dictionary<string, string> GetVersionsFromResult(Provider results)
+        {
+            Dictionary<string, string> resourceVersions = new Dictionary<string, string>();
+            foreach (var type in results.Data.ResourceTypes)
+            {
+                resourceVersions[type.ResourceType] = type.ApiVersions[0];
+            }
+            return resourceVersions;
+        }
+
+        internal string GetApiVersionForNamespace(string resourceNamespace, CancellationToken cancellationToken = default)
+        {
+            string version;
+            if (!ClientOptions.NamespaceVersions.TryGetValue(resourceNamespace, out version))
```

Comment:
NamespaceVersions is not used outside this type, should we move it onto ProviderCollection?

Reply:
Same answer as above